src/looprestoration.rs

```diff
@@ -1556,32 +1556,1851 @@ unsafe fn sgr_filter_3x3_neon<BD: BitDepth>(
     );
 }
 
+#[cfg(all(feature = "asm", target_arch = "aarch64"))]
+unsafe fn rotate5_x2(sumsq_ptrs: &mut [*mut i32; 5], sum_ptrs: &mut [*mut i16; 5]) {
+    sumsq_ptrs.rotate_left(2);
+    sum_ptrs.rotate_left(2);
+}
+
+#[cfg(all(feature = "asm", target_arch = "aarch64"))]
+unsafe fn rotate_ab_2(A_ptrs: &mut [*mut i32; 2], B_ptrs: &mut [*mut i16; 2]) {
+    A_ptrs.rotate_left(1);
+    B_ptrs.rotate_left(1);
+}
+
+#[cfg(all(feature = "asm", target_arch = "aarch64"))]
+unsafe fn rotate_ab_3(A_ptrs: &mut [*mut i32; 3], B_ptrs: &mut [*mut i16; 3]) {
+    A_ptrs.rotate_left(1);
+    B_ptrs.rotate_left(1);
+}
+
+#[cfg(all(feature = "asm", target_arch = "aarch64"))]
+unsafe fn rotate_ab_4(A_ptrs: &mut [*mut i32; 4], B_ptrs: &mut [*mut i16; 4]) {
+    A_ptrs.rotate_left(1);
+    B_ptrs.rotate_left(1);
+}
+
 #[cfg(all(feature = "asm", any(target_arch = "arm", target_arch = "aarch64")))]
-unsafe extern "C" fn sgr_filter_mix_neon_erased<BD: BitDepth>(
-    p: *mut DynPixel,
-    stride: ptrdiff_t,
-    left: *const LeftPixelRow<DynPixel>,
-    lpf: *const DynPixel,
+unsafe fn rav1d_sgr_box3_row_h_neon<BD: BitDepth>(
+    sumsq: *mut i32,
+    sum: *mut i16,
+    left: *const [BD::Pixel; 4],
+    src: *const BD::Pixel,
     w: c_int,
-    h: c_int,
-    params: *const LooprestorationParams,
     edges: LrEdgeFlags,
-    bitdepth_max: c_int,
+    bd: BD,
 ) {
-    sgr_filter_mix_neon(
-        p.cast(),
-        stride,
+    macro_rules! asm_fn {
+        ($name:ident) => {{
+            extern "C" {
+                fn $name(
+                    sumsq: *mut i32,
+                    sum: *mut i16,
+                    left: *const c_void,
+                    src: *const c_void,
+                    w: c_int,
+                    edges: LrEdgeFlags,
+                    bitdepth_max: c_int,
+                );
+            }
+            $name
+        }};
+    }
+    (match BD::BPC {
+        BPC::BPC8 => asm_fn!(dav1d_sgr_box3_row_h_8bpc_neon),
+        BPC::BPC16 => asm_fn!(dav1d_sgr_box3_row_h_16bpc_neon),
+    })(sumsq, sum, left.cast(), src.cast(), w, edges, bd.into_c())
```

Comment:
```suggestion
    bd_fn!(asm_fn, BD, sgr_box3_row_h, neon)(sumsq, sum, left.cast(), src.cast(), w, edges, bd.into_c())
```

You can use `bd_fn!` for these (same applies for all the other patterns like this), and doing so makes fixing #614 simpler.